keps/sig-apps/0008-sidecarcontainer.md

```diff
@@ -0,0 +1,145 @@
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints-optional)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Alternatives](#alternatives-optional)
+
+## Summary
+
+To solve the problem of container lifecycle dependency we can create a new class of container: a "sidecar container" that behaves primarily like a normal container but is handled differently during termination and startup.
+
+## Motivation
+
+SideCar containers have always been used in some ways but just not formally identified as such, they are becoming more common in a lot of applications and as more people have used them, more issues have cropped up.
+
+Here are some examples of the main problems:
+
+### Jobs
+ If you have a Job with two containers one of which is actually doing the main processing of the job and the other is just facilitating it, you encounter a problem when the main process finishes; your sidecar container will carry on running so the job will never finish.
+
+The only way around this problem is to manage the sidecar container's lifecycle manually and arrange for it to exit when the main container exits. This is typically achieved by building an ad-hoc signalling mechanism to communicate completion status between containers. Common implementations use a shared scratch volume mounted into all pods, where lifecycle status can be communicated by creating and watching for the presence of files. This pattern has several disadvantages:
+
+* Repetitive lifecycle logic must be rewritten in each instance a sidecar is deployed.
+* Third-party containers typically require a wrapper to add this behaviour, normally provided via an entrypoint wrapper script implemented in the k8s container spec. This adds undesirable overhead and introduces repetition between the k8s and upstream container image specs.
+* The wrapping typically requires the presence of a shell in the container image, so this pattern does not work for minimal containers which ship without a toolchain.
+
+### Startup
+An application that has a proxy container acting as a sidecar may fail when it starts up as it's unable to communicate until its proxy has started up successfully. Readiness probes don't help if the application is trying to talk outbound.
+
+### Shutdown
+Applications that rely on sidecars may experience a high amount of errors when shutting down as the sidecar may terminate before the application has finished what it's doing.
+
+
+## Goals
+
+Solve issues so that they don't require application modification:
+[25908](https://github.com/kubernetes/kubernetes/issues/25908) - Job completion
+[65502](https://github.com/kubernetes/kubernetes/issues/65502) - Container startup dependencies
+
+## Non-Goals
+
+Allowing multiple containers to run at once during the init phase.
```

Comment:
In cases when you send outbound traffic through a proxy sidecar, it seems reasonable that you may want to have an init container that makes outbound calls as part of setting up the Pod environment. A user gave an example [here](https://github.com/kubernetes/kubernetes/issues/25908#issuecomment-404282114) involving the Cloud SQL Proxy.

On the other hand, we know that in other cases sidecars need to run *after* initContainers because they depend on the results (e.g. setting up iptables before starting Envoy).

If we're trying to come up with a fully fleshed proposal for how to facilitate the sidecar pattern, I think we should at least consider both cases. It might be that the answer is, "we can't think of a clean way to handle both right now", but I haven't seen evidence yet that we have tried and failed to do that, so I wouldn't call it a non-goal just yet.

As a strawman (not proposing we do exactly this; just saying it's not inconceivable), you could replace the boolean `sidecar` field with a decomposed set of fields to control behavior in orthogonal dimensions:

```yaml
containers:
- name: ...
  # PreInit means start during Initialization phase and wait until it's Ready before running the first initContainer.
  # PostInit means start after the last initContainer finishes and wait until it's Ready before starting any Unordered containers.
  # Unordered is the current behavior (default value).
  startupOrder: (PreInit|PostInit|Unordered)

  # Deferred means wait until all Unordered containers have terminated before terminating this.
  #   Also, if all non-deferred containers have reached a terminal state
  #   (they already ran and they're not going to be restarted by policy),
  #   then start terminating all deferred containers.
  # Unordered is the current behavior (default value).
  shutdownOrder: (Deferred|Unordered)
- ...
```

Reply:
@enisoc it seems that we're just abbreviating the configuration for "container dependencies", which is fine. But I'm not sure we would need to separate startup/shutdown ordering. I think the startup order implies the shutdown order.

If you have container A that should start before container B, then you're effectively saying that B *depends on* A.  So the lifecycles would look like:

```
|------------ A ------------|
    |--------- B --------|

time ->
```
This conceptually works regardless of whether B is an init container which runs to completion (requiring sidecar A).

I wonder if we could capture this in a single property.  Something like this?

``` yaml
containers:
- name: ...
  lifecycleOrder:
    # PreInit: start during Initialization phase (but before Init containers). Also end after all
    # other containers have been terminated.
    #
    # Init: (the default for init containers) start during the Initialization phase, after
    # PreInit containers are Ready. These will generally run to completion, but shutdown
    # order will be after PostInit, but before PreInit.
    #
    # PostInit: start after the last Init container finishes and wait until it's ready.
    # Shutdown after default containers but before Init containers.
    #
    # Default: (default for non-init containers) start after all other containers are Ready.
    # Shutdown before all other containers.
    phase: (preInit|Init|PostInit|Default)

    # Not needed? This might enable us to break ties when we have
    # 2 containers in the same phase where one depends on the other.
    dependsOn: (containerName)
```

Reply:
Currently the init phase and the running phase are very clearly separate, creating a container that would run through both is quite a big change (not inconceivable). 

Just to suggest an alternative, if the desire is to just have sidecars during the init phase you could instead just allow something like `sidecar: true` in the initContainer spec, these init sidecars would start before normal init and exit when all the init containers have completed. So very similar to what's being suggested just in the init phase instead.

Reply:
That seems reasonable. @enisoc WDYT?